rclcpp/src/rclcpp/detail/mutex_two_priorities.cpp

```diff
@@ -31,13 +31,34 @@ HighPriorityLockable::HighPriorityLockable(MutexTwoPriorities & parent)
 void
 HighPriorityLockable::lock()
 {
-  parent_.data_.lock();
+  bool did_aument_count{false};
+  std::unique_lock<std::mutex> guard{parent_.cv_mutex_};
+  if (parent_.data_taken_) {
+    ++parent_.hp_waiting_count_;
+    did_aument_count = true;
+  }
+  while (parent_.data_taken_) {
+    parent_.cv_.wait(guard);
+  }
+  if (did_aument_count) {
+    --parent_.hp_waiting_count_;
+  }
+  parent_.data_taken_ = true;
 }
 
 void
 HighPriorityLockable::unlock()
 {
-  parent_.data_.unlock();
+  {
+    std::lock_guard<std::mutex> guard{parent_.cv_mutex_};
+    parent_.data_taken_ = false;
+  }
+  // We need to notify_all(), and not only one,
+  // because the notified thread might be low priority
+  // when a high priority thread is still waiting.
+  // In that case, the low priority thread will go to sleep again,
+  // release the mutex, and the high priority thread can continue.
+  parent_.cv_.notify_all();
```

Comment:
@ivanpauno I know this is tricky code, and that you've carefully thought this true. Having said that, wouldn't having one condition variable per priority, and notifying one on each, boost efficiency a bit?

Reply:
Yes, I thought about that when writting the comment as well, but it was late and I didn't have much brain power left :smile: .
I will give it a try now.

Reply:
Done in https://github.com/ros2/rclcpp/pull/1628/commits/909bca576ac3b2d4b3d8bd108efc21eda42ca618.
I'm almost sure the code is still correct, though the version with one conditional variable is more trivial to follow.
PTAL